### ISSUE TITLE
Seamless Decentralized Nano Node Sharding and Fail Safe

### ISSUE THREAD

**Issue by JxJay42** [nanocurrency/nano-node#4343]
## Description
This proposal aims to enhance the Nano network by introducing a system that allows for:
- Sharding of Nano nodes/representatives without a central network manager or coordinator.
- Easy addition of new shards without requiring configuration changes in existing shards.
- An automatic fail-safe mechanism to ensure continuous operation, where if a shard becomes non-functional another shard votes on accounts formerly assigned to the lost shard.
- Flexibility in shard locations, eliminating the need for VPNs unless specified in configurations.

## Requirements:
- Peer authorization.
- Ability for multiple nodes on the network using the same private key.
- Public IP/connectivity of nodes. This can be theoreticaly solved by some network messaging but is more complicated, or can be solved if the non public shards are on the same network or VPN.

## Concept Summary
Nodes declare their "shard order" during peering. This order number is used as a base for an account address binary mask, defining whether that account's blocks will be processed (voted on) by this node. 
For example, a node with shard order 0 processes accounts ending in bit 0, and one with order 1 processes accounts ending in bit 1.
All nodes receive all final blocks for database consistency.

### Terminology:
S_Node: A group of shards under the same private key/account of a representative.
S_Order: Shard order number of a specific shard.
account_space: The binary segment of accounts managed by a shard. xxx0 means all accounts ending in bit 0.
account_mask: A binary representation of a shard's responsibility for certain account spaces. Mask 0001 for shard 0 means: account_mask & S_Order = 0001 & 0000 = 0000 = responsibility for xxx0 space

### Peering Process

Nodes declare their S_Order during peering.

For any node that is not a part of S_Node, the S_Order parameter is almost completely irrelevant. The only use it can have is to possibly identify wrong network configuration or possibly malicious nodes.

S_Order is crucial for shards within the same S_Node, dictating which account spaces they manage.
To decide which part of account's space to process and vote on, each shard should know:
- All other available shards
- Whether these shards are online
- The other shard's S_Orders

The easiest method is that every shard has to be connected to all other shards. But this comes with some possible overhead on large shardings, and some double votings if connections get dropped without the shards going down (failsafe feature of this proposal).

The more complicated method, is that shards message all connected shards about their connection list with connected S_Orders. This would manage consistency even if some connections are severed, without that shard going down. But this could have delays in reaction times on voting failsafe (depends on messaging implementation). 

Why delays?
If a shard only builds its account_mask by using what is connected to it directly, then the reaction time for failsafe would be micro seconds. Because if any network change happens, it just recalculates account_mask. 
With messaging it is more complicated, because we can not have shards sending their connected list immediately after each network change. That would just completely wreck the network on unstable connections. So we would need a delay, so that there is a chance to reconnect lost shards with a bad connection. Both methods could be combined, but on complicated networks it will create delays.

### Shard order and account mask

The fail-safe feature depends on computing the correct account mask based on other available shards.
Let's have 4 shards:
| S_order | binary | account_space | account_mask |
|--|--|--|--|
| 0 | 00 | xxxxxx00 | 00000011 |
| 1 | 01 | xxxxxx01 | 00000011 |
| 2 | 10 | xxxxxx10 | 00000011 |
| 3 | 11 | xxxxxx11 | 00000011 |

Such configuration will in parallel process the whole account space.
What happens if S1 drops?
Shards recompute their account masks when they detect a change in the network, such as a shard dropping off.
Algorithmically :
- Shards 2 and 3 still complement each other for the account_space xx1x, so they are not affected by what happens in account_space xx0x.
- Therefore only shard0 is affected.
- It manages xxxxxx00.
- Does some other shard manage last bit switched (xxxxxx01)?
- No.
- Account_space changes to xxxxxx0x, which means that shard0 will now cover also accounts of the dropped shard. It does so by changing the account_mask to 00000010.
- Does some other shard manage 2nd bit switched (xxxxxx1x)?
- Yes.
- End of search, final account_space is xxxxxx0x.

This means that shard numbers don't need to be incremental.
Let's have two shards: S1 and S14
Account masks will be computed : 
| S_order | binary | account_space | account_mask |
|--|--|--|--|
| 1 | 0001 | xxxx0xxx | 00001000 |
| 14 | 1110 | xxxx1xxx |  00001000 |

The number of shards (S_Order) should be limited to avoid network complexity and potential issues.
Although it is mathematically possible to have an S_Order as high as the max account number, this would only result in chaos and too long account mask search.
Initially a max S_Order of 15 or 31 would be more than enough. Even 8 shards for the first version would be enough, so that the network gradually starts using the feature.
Too high a number of shards could increase chances for a split brain, until a better full network messaging is implemented. 

### Account mask 

Account mask enables comparison of account number and S_Order.

With this binary comparison:
```
if( account_number & account_mask == s_order & account_mask){
  //process
}
```
Let's say the account's last byte looks like 10110111
binary math "AND" with account mask :
`10110111 & 00001000 = 00000000`
With the shard table
| S_order | binary | process account mask | s_order & mask |
|--|--|--|--|
| 1 | 0001 | 00001000 | 00000000 |
| 14 | 1110 | 00001000 | 00001000 |
Shard with number 1 will process that account.

### Account mask computation from shard list.
Some S_Order combinations create weird situations.
Example of it are shards [0, 1, 8].
This creates gaps in account masks, and the mask for shard0 is 00001001 .

Explaining the computation algorithm I found in plain text would be long and convoluted.

Instead I wrote a simple python script.
It is supposed to be explicit in logic, so it is definitely not the most effective and not "pythonic". But it works.

Example result for a list of shards 0, 1, 8, 10, 12

|s_order | bin      | a_mask
|--|--|--|
|    000 | 00000000 | 00001001
|    001 | 00000001 | 00001001
|    008 | 00001000 | 00001110
|    010 | 00001010 | 00001110
|    012 | 00001100 | 00001100

```
def find_account_mask(current, shards, mask_bit_length):
    # account_mask
    a_mask = 0
    # moving from lowest bit to max used bit to check for group responsibility
    for bit in range(0, mask_bit_length):
        # masking binary part to compare
        current_mask = ~((1 << bit) - 1)
        # lowest tested bit flipped
        flip_bit = ~current & (1 << bit)
        # flip the last significant bit in the current s_order
        current_bit_flip = (current & (current_mask << 1)) | flip_bit
        # searching, if there is a paired shard for current mask group
        for s_order in shards:
            if s_order == current:
                continue
            s_masked = s_order & current_mask
            if s_masked == current_bit_flip:
                # if this shard has a pair, for this mask group,
                # this bit will be significant for the account mask of this shard
                a_mask |= 1 << bit
                break
    return a_mask


shards = [
    0, 1, 8, 10, 12,
]

shards.sort(reverse=True)
mask_bit_length = shards[0].bit_length()
shards.sort(reverse=False)

# compute account mask for all shards
rows = []
for s_order in shards:
    a_mask = find_account_mask(s_order, shards, mask_bit_length)
    row = {
        's_order': f"{s_order:03d}",
        'bin': f"{s_order:08b}",
        'a_mask': f"{a_mask:08b}"
    }
    rows.append(row)

# print a table
print('s_order', 'bin', 'a_mask', sep=" | ")
for row in rows:
    print("    " + row['s_order'], row['bin'], row['a_mask'], sep=" | ")
```

### Flow of operations
It is highly possible that I'm not familiar enough with current node protocols, and that this part is wrong or misleading.
I rely on the dev team to infer the correct implementation from my intentions.

1 . During peering or during relevant messaging, shard will compute account_mask, which must be for subsequent parts together with S_Order.
2 . Block processing into backlog, should not be changed.
3 . During filling of buckets, the node should only take blocks which satisfy the condition `(account_number & account_mask) == (s_order & account_mask)`
4 . No changes in AEC
5 . If a final block is received of a block that is waiting in backlog, it should be removed from backlog - but this should be already the current functionality.

### Delay between peering and voting
A node that connects to the network, is most probably not yet ready to start voting - maybe bootstrapping.
This has to be messaged to other shards either sending S_Order NULL, or a flag signalling active shard.

### Voting weight

Because of fail-safe feature of this sharding, voting weight does not have to be split to some subnetworks.
Exactly like now, if at least one node is online, its voting weight is added to the network.

### Other possibly affected areas

I don't know how this could affect vote hinting, vote storage and similar.
Bootstrapping should not be affected.
Maybe there is a special situation, if one shard is dropped, between first and final vote on a block.

### CPS/TPS
The most obvious benefit is the multiplication of max achievable CPS, as operations are parallelized on different hardware and network.
This multiplication can be done with multiple low powered nodes rather than one powerful node.

### Fail safe
Automatic account mask configuration makes sharded nodes almost impossible to take down.
This also gives benefits for admins, for rolling node upgrades without the fear of affecting the nano network.

### Distributed
As the only requirement is to have a public IP, so that shards can be connected to each other directly, the geographical location of shards is unlimited. It will be impossible for any big actor to take down the nano network by blocking some big datacenter.

### Decentralized
No governor and no single point of failure are the base ideals of cryptocurrency.

### Multiple nodes for apps api access
Applications use nodes to access the network, and for self-sufficiency they use their own nodes.
This will allow to have multiple nodes under the same key and have fail safes for the aplication, for uninterrupted service.

### Potential imbalances in account space distribution among shards.
Certain combinations of S_Orders can create imbalances in terms of how large an account space each shard processes.
Lets add Shard4 to the first example:
| S_order | binary | account_space | assigned account space ratio |
|--|--|--|--|
| 0 | 000 | xxxxx000 | 12.5% |
| 1 | 001 | xxxxx001 | 12.5% |
| 2 | 010 | xxxxx010 | 12.5% |
| 3 | 011 | xxxxx011 | 12.5% |
| 4 | 100 | xxxxx1xx | 50% |

But such imbalances are completely in the hands of the admin, and can be also desired - i.e. it could be statistically inferred that some account space has bigger load.

### Split brain situations
Imagine a structure where a node is on multiple shards in two different locations.
Those locations lose connection with each other (i.e. unintended fw block), but both are still connected to the nano network.
Both parts will automatically reconfigure themselves to manage the whole account space, and both will vote simultaneously.
This should not be a big problem, if one private key is voting on the same block with multiple nodes votes are counted by signing, and the same signature is counted only once.
These situations are only solvable by network wide messaging, which could be complicated and is not part of this proposal.

### Handling of S_Order conflicts and the need for administrative intervention.
Two shards could declare the same S_Order.
I don't think that there should be any "automatic" change of S_Order. This situation should be brought to the attention of the administrator, and one of the nodes should stop voting on blocks.
The resolution could be as simple as declaring uptime, and the older one wins.
Even public tools, that show the status of nodes (like nanolooker) could show such misconfigurations.

### Possible misuse and attacks

I was unable to find any possible misuse.
Peer authentication solves most of the possible issues.
Multiple nodes for the same key can already be created now.
S_Order messaging would only affect the S_Node it self.

## Addendum
Special thanks to Senatus for edits of this proposal, that it better follows the rules of English language.

### What problem would be solved by this feature?

Performance, uptime, reliability.

### Are there any previous requests for this feature?

Unknown

### Do you have a suggested solution?

Described in the text

### If this feature is approved, would you be willing to submit a pull request with the solution?

I would not

**Comment by JxJay42** [comment]
## Aditional thoughts.

### Bootstrap
Shards could prioritize their own node to bootstrap.
Or even better, directly download a signed database from another shard.
Many posibilities.
Base thought is, shard can trust another signed shard in its own S_Node.
This could even encourage admins, to hold one active low power non-voting node just for bootstrapping new shards. Code could even recognize such a configuration, and prioritly sync bootstrap from non-voting nodes which are part of S_Node.

### Network communication optimisations
Other global nodes could optimize unchecked block propagation.
I.e. unchecked blocks could be send only to one shard of the S_Node, random one.
The responsibility would then fall onto the shard, to distribute unchecked blocks among its sisters.
That could also be optimized, that the block is send only to two shards which are closest to account_space responsibility of that block. (two nodes, as a fail safe).
